File: frontend/src/app/chat/services/symmetric-crypto.ts
// AES-GCM symmetric encryption for chat messages
// Usage: await SymmetricCrypto.encrypt(plaintext, key) and await SymmetricCrypto.decrypt(ciphertext, iv, key)
export class SymmetricCrypto {
  static async generateKey(): Promise<string> {
    const key = await window.crypto.subtle.generateKey(
      { name: 'AES-GCM', length: 256 },
      true,
      ['encrypt', 'decrypt']
    );
    const raw = await window.crypto.subtle.exportKey('raw', key);
    return btoa(String.fromCharCode(...new Uint8Array(raw)));
  }

  static async encrypt(plaintext: string, keyB64: string): Promise<{ ciphertext: string; iv: string }> {
    const key = await window.crypto.subtle.importKey(
      'raw',
      Uint8Array.from(atob(keyB64), c => c.charCodeAt(0)),
      'AES-GCM',
      false,
      ['encrypt']
    );
    const iv = window.crypto.getRandomValues(new Uint8Array(12));
    const enc = new TextEncoder();
    const ciphertext = await window.crypto.subtle.encrypt(
      { name: 'AES-GCM', iv },
      key,
      enc.encode(plaintext)
    );
    return {
      ciphertext: btoa(String.fromCharCode(...new Uint8Array(ciphertext))),
      iv: btoa(String.fromCharCode(...iv))
    };
  }

  static async decrypt(ciphertextB64: string, ivB64: string, keyB64: string): Promise<string> {
    const key = await window.crypto.subtle.importKey(
      'raw',
      Uint8Array.from(atob(keyB64), c => c.charCodeAt(0)),
      'AES-GCM',
      false,
      ['decrypt']
    );
    const iv = Uint8Array.from(atob(ivB64), c => c.charCodeAt(0));
    const ciphertext = Uint8Array.from(atob(ciphertextB64), c => c.charCodeAt(0));
    const decrypted = await window.crypto.subtle.decrypt(
      { name: 'AES-GCM', iv },
      key,
      ciphertext
    );
    return new TextDecoder().decode(decrypted);
  }
}

File: backend/app/core/symmetric_crypto.py
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
import os

class SymmetricCrypto:
    @staticmethod
    def generate_key() -> str:
        key = AESGCM.generate_key(bit_length=256)
        return base64.b64encode(key).decode()

    @staticmethod
    def encrypt(plaintext: bytes, key_b64: str) -> dict:
        key = base64.b64decode(key_b64)
        iv = os.urandom(12)
        aesgcm = AESGCM(key)
        ciphertext = aesgcm.encrypt(iv, plaintext, None)
        return {
            'iv': base64.b64encode(iv).decode(),
            'ciphertext': base64.b64encode(ciphertext).decode()
        }

    @staticmethod
    def decrypt(ciphertext_b64: str, iv_b64: str, key_b64: str) -> bytes:
        key = base64.b64decode(key_b64)
        iv = base64.b64decode(iv_b64)
        ciphertext = base64.b64decode(ciphertext_b64)
        aesgcm = AESGCM(key)
        return aesgcm.decrypt(iv, ciphertext, None)
